util/src/network/host.rs

```diff
@@ -662,17 +637,37 @@ impl<Message> Host<Message> where Message: Send + Sync + Clone {
 			let mut s = session.lock().unwrap();
 			match s.readable(io, &self.info.read().unwrap()) {
 				Err(e) => {
-					trace!(target: "network", "Session read error: {}:{} ({:?}) {:?}", token, s.id(), s.remote_addr(), e);
+					trace!(target: "network", "Session read error: {}:{:?} ({:?}) {:?}", token, s.id(), s.remote_addr(), e);
 					match e {
 						UtilError::Network(NetworkError::Disconnect(DisconnectReason::UselessPeer)) |
 						UtilError::Network(NetworkError::Disconnect(DisconnectReason::IncompatibleProtocol)) => {
-							self.nodes.write().unwrap().mark_as_useless(s.id());
+							if let Some(id) = s.id() {
+								self.nodes.write().unwrap().mark_as_useless(id);
+							}
 						}
 						_ => (),
 					}
 					kill = true;
 				},
 				Ok(SessionData::Ready) => {
+					if !s.info.originated {
+						let session_count = self.session_count();
+						let ideal_peers = { self.info.read().unwrap().deref().config.ideal_peers };
+						if session_count >= ideal_peers as usize {
+							s.disconnect(DisconnectReason::TooManyPeers);
+							return;
+						}
+						// Add it no node table
+						if let Ok(address) = s.remote_addr() {
+							let entry = NodeEntry { id: s.id().unwrap().clone(), endpoint: NodeEndpoint { address: address, udp_port: address.port() } };
+							self.nodes.write().unwrap().add_node(Node::new(entry.id.clone(), entry.endpoint.clone()));
+							let mut discovery = self.discovery.lock().unwrap();
+							if let Some(ref mut discovery) = *discovery.deref_mut() {
+								discovery.add_node(entry);
+							}
+						}
+					}
+					self.num_sessions.fetch_add(1, AtomicOrdering::Relaxed);
```

Comment:
I wonder if the `Relaxed` ordering would lead to potentially incorrect values being returned from `session_count`. This doesn't pose a problem on the x86 or x86_64 memory model, so any bugs introduced from `Relaxed` atomics could be hidden from most of us.

The performance gains from `Relaxed` vs `SeqCst` are not that significant, so why not trend towards the safe side?